blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md

```diff
@@ -0,0 +1,239 @@
+cases and subgoals, and then I refine them to add more automation for
+robustness. I found bullets very useful for the first step, and during
+the refinement process.
+
+# Bullets
+
+Bullets are actually a combination of two features, braces `{ ... }`
+and actually list bullets, `-`, `+`, or `*` or homogeneous repetitions
+of those, for example `--` or `***`.
+
+## Braces
+
+The opening brace `{` focuses the proof on the current subgoal. If you
+finish the proof of the subgoal, the following subgoal will not become
+accessible automatically; you have to use the closing brace `}`
+first. (If you finish the goal earlier than you think, you get an
+error.) Conversely, `}` fails if the subgoal is not complete. (If you
+fail to finish, you get an error.)
+
+The previous example can thus be written as follows, and will be more robust:
+
+```
+induction foo. (* this creates many subgoal *)
+{
+  proof of first subgoal.
+}
+{
+  proof of second subgoal.
+}
+```
+
+If you also want to make sure that an error occurs if the number of
+subgoals changes (for example if new constructors are added to the
+inductive type of `foo`), you can use an outer layer of braces:
+
+
+```
+{ induction foo. (* this creates many subgoal *)
+  {
+    proof of first subgoal.
+  }
+  {
+    proof of second subgoal.
+  }
+} (* would fail if a new subgoal appeared *)
+```
+
+## List bullets
+
+A bullet, for example `--`, also focuses on the next subgoal. The
+difference is that when the subgoal is finished, you do not have
+a closing construction, you must use the same bullet to move to the
+next subgoal. (Again, this fails if the first proof step changes to
+prove too much or too little.) With bullets you would write
+
+```
+induction foo. (* this creates many subgoal *)
++ proof of first subgoal.
++ proof of second subgoal.
+```
+
+Bullets can be nested, but you must use different bullets for the
+different nesting levels. For example, if this proof is only one
+subgoal of a larger proof, you can use:
+
+```
+- induction foo. (* this creates many subgoal *)
+  + proof of first subgoal.
+  + proof of second subgoal.
+- (* would fail if a new subgoal appeared *)
+  rest of the proof
+```
+
+The natural ordering of bullets, I think, is by increasing number of
+lines: `-`, `+` then `*` (and then multi-character bullets,
+I guess). You can also mix bullets with braces: the opening brace
+resets the bullet scope, any bullet can be used again with the
+subgoal.
+
+This gives a large space of freedom in how you want to use these
+features. You can use only braces, only bullets, braces and only one
+level of bullets, etc. My own style evolved with experience using the
+feature, and I will present the current status below.
+
+# My current bullet style
+
+When deciding how to use bullets, one distinguishes the commands that
+preserve the number of subgoals and those that may create new
+subgoals. I do some further distinctions.
```

Comment:
"I use some additional distinctions" or "I employ some additional distinctions" would be better here